libnetwork/resolver.go

```diff
@@ -479,19 +479,6 @@ func (r *Resolver) forwardExtDNS(proto string, query *dns.Msg) *dns.Msg {
 			// Server returned REFUSED: this can be a transitional status, so continue with the next external DNS server
 			logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), queryName)
 			continue
-		case dns.RcodeNameError:
-			// Server returned NXDOMAIN. Stop resolution if it's an authoritative answer (see RFC 8020: https://tools.ietf.org/html/rfc8020#section-2)
-			logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), queryName)
-			if resp.Authoritative {
-				break
-			}
-			continue
```

Comment:
Still somewhat on the fence on this, as this would be a breaking change (even if the "split horizon" scenario is wrong?). 

TBH part of https://datatracker.ietf.org/doc/html/rfc1035#page-27 is a bit ambiguous (or at least, I have no idea how it should be interpreted) (what does "meaningful only" mean? Should be ignored otherwise?);

```
3               Name Error - Meaningful only for
                responses from an authoritative name
                server, this code signifies that the
                domain name referenced in the query does
                not exist.
```

Reply:
As far as I can tell, "meaningful" is in reference to recursive DNS resolvers, as in a recursive resolver should treat non-authoritative NXDOMAIN responses as a signal to continue recursing to the authoritative name server to get an authoritative response. We are a stub resolver, not a recursive resolver, so our duty is to faithfully return the response provided by the recursive resolver we delegated the handling of the query onto.